domain/secretbackend/state/state.go

```diff
@@ -746,6 +756,43 @@ WHERE b.%s = $M.identifier`, columName)
 	return rows.toSecretBackends()[0], nil
 }
 
+// GetActiveModelSecretBackend returns the active secret backend ID and config for the given model.
+// It returns an error satisfying [modelerrors.NotFound] if the model provided does not exist.
+func (s *State) GetActiveModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID) (string, *provider.ModelBackendConfig, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", nil, internalerrors.Capture(err)
+	}
+	var (
+		modelBackend secretbackend.ModelSecretBackend
+		backend      *secretbackend.SecretBackend
+	)
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		modelBackend, err = s.getModelSecretBackendDetails(ctx, tx, modelUUID)
+		if err != nil {
+			return internalerrors.Capture(err)
+		}
+		if modelBackend.ModelType == coremodel.CAAS && modelBackend.SecretBackendName == kubernetes.BackendName {
+			backend, err = s.getK8sSecretBackendForModel(ctx, tx, modelUUID)
+		} else {
+			backend, err = s.getSecretBackend(ctx, tx, secretbackend.BackendIdentifier{ID: modelBackend.SecretBackendID})
+		}
```

Comment:
Now that we support proper k8s backend definitions, we really should rework this (soon) so that this special case code can be removed.

Reply:
Agreed.